### PR TITLE
Define simple XPath based abstract check

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: xml-parsing/src/main/java/org/sonarsource/analyzer/commons/xml/checks/SimpleXPathBasedCheck.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+import org.sonar.api.utils.log.Loggers;
+import org.w3c.dom.NamedNodeMap;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+
+public abstract class SimpleXPathBasedCheck extends SonarXmlCheck {
+
+  private static final Logger LOG = Loggers.get(SimpleXPathBasedCheck.class);
+
+  private final XPath xpath = XPathFactory.newInstance().newXPath();
```
Comment (vilchik-elena) [inline xml-parsing/src/main/java/org/sonarsource/analyzer/commons/xml/checks/SimpleXPathBasedCheck.java]: can we use constructor instead?

File: xml-parsing/src/main/java/org/sonarsource/analyzer/commons/xml/checks/SimpleXPathBasedCheck.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  public XPathExpression getXPathExpression(String expression) {
+    try {
+      return xpath.compile(expression);
+    } catch (XPathExpressionException e) {
+      throw new IllegalStateException(String.format("[%s] Fail to compile XPath expression '%s'.", ruleKey(), expression), e);
+    }
+  }
+
+  @CheckForNull
+  public NodeList evaluate(XPathExpression expression, Node node) {
```
Comment (vilchik-elena) [inline xml-parsing/src/main/java/org/sonarsource/analyzer/commons/xml/checks/SimpleXPathBasedCheck.java]: how about adding a doc explaining when result is null?

File: xml-parsing/src/main/java/org/sonarsource/analyzer/commons/xml/checks/SimpleXPathBasedCheck.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+
+public abstract class SimpleXPathBasedCheck extends SonarXmlCheck {
+
+  private static final Logger LOG = Loggers.get(SimpleXPathBasedCheck.class);
+
+  private final XPath xpath = XPathFactory.newInstance().newXPath();
+
+  public XPathExpression getXPathExpression(String expression) {
```
Comment (vilchik-elena) [inline xml-parsing/src/main/java/org/sonarsource/analyzer/commons/xml/checks/SimpleXPathBasedCheck.java]: how about adding a doc explaining when exception is thrown by the method?

File: xml-parsing/src/main/java/org/sonarsource/analyzer/commons/xml/checks/SimpleXPathBasedCheck.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    if (numberResults == 0) {
+      return Collections.emptyList();
+    }
+    return IntStream.range(0, numberResults)
+      .mapToObj(nodeList::item)
+      .collect(Collectors.toList());
+  }
+
+  @CheckForNull
+  public static Node nodeAttribute(Node node, String attribute) {
```
Comment (vilchik-elena) [inline xml-parsing/src/main/java/org/sonarsource/analyzer/commons/xml/checks/SimpleXPathBasedCheck.java]: I suggest to move this one to `XmlFile` class

File: xml-parsing/src/main/java/org/sonarsource/analyzer/commons/xml/checks/SimpleXPathBasedCheck.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+  @CheckForNull
+  public NodeList evaluate(XPathExpression expression, Node node) {
+    try {
+      return (NodeList) expression.evaluate(node, XPathConstants.NODESET);
+    } catch (XPathExpressionException e) {
+      if (LOG.isDebugEnabled()) {
+        RuleKey ruleKey = ruleKey();
+        LOG.debug(String.format("[%s] Unable to evaluate XPath expression on file %s", ruleKey, inputFile()));
+        LOG.error(String.format("[%s] XPath exception:", ruleKey), e);
```
Comment (vilchik-elena) [inline xml-parsing/src/main/java/org/sonarsource/analyzer/commons/xml/checks/SimpleXPathBasedCheck.java]: how about simplifying this to:
```
if (LOG.isDebugEnabled()) {
    RuleKey ruleKey = ruleKey();
    LOG.error(String.format("[%s] Unable to evaluate XPath expression on file %s", ruleKey, inputFile()), e);
}
```

File: xml-parsing/src/test/java/org/sonarsource/analyzer/commons/xml/checks/SimpleXPathBasedCheckTest.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    SimpleXPathBasedCheck check = new SimpleXPathBasedCheck() {
+      XPathExpression bs = getXPathExpression("//b");
+      XPathExpression ds = getXPathExpression("//d");
+
+      @Override
+      protected void scanFile(XmlFile file) {
+        Document node = file.getDocument();
+
+        NodeList result = evaluate(bs, node);
+        assertThat(result).isNotNull();
```
Comment (vilchik-elena) [inline xml-parsing/src/test/java/org/sonarsource/analyzer/commons/xml/checks/SimpleXPathBasedCheckTest.java]: is there a way to assert presence of assertions in the test? I think it's useful when assert is inside nested call like here.

File: xml-parsing/src/test/java/org/sonarsource/analyzer/commons/xml/checks/SimpleXPathBasedCheckTest.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+        assertThat(empty.getLength()).isEqualTo(0);
+        List<Node> emptyAsList = evaluateAsList(ds, node);
+        assertThat(emptyAsList).isEmpty();
+
+        assertThat(asList(null)).isEmpty();
+      }
+    };
+
+    check.scanFile(null, RuleKey.of("tst", "nodeList"), xmlFile);
+    assertThat(logTester.logs()).isEmpty();
```
Comment (vilchik-elena) [inline xml-parsing/src/test/java/org/sonarsource/analyzer/commons/xml/checks/SimpleXPathBasedCheckTest.java]: should we set logs to debug before?

File: xml-parsing/src/test/java/org/sonarsource/analyzer/commons/xml/checks/SimpleXPathBasedCheckTest.java [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+        List<Node> asList = evaluateAsList(bs, node);
+        assertThat(asList).hasSize(4);
+
+        NodeList empty = evaluate(ds, node);
+        assertThat(empty).isNotNull();
+        assertThat(empty.getLength()).isEqualTo(0);
+        List<Node> emptyAsList = evaluateAsList(ds, node);
+        assertThat(emptyAsList).isEmpty();
+
+        assertThat(asList(null)).isEmpty();
```
Comment (vilchik-elena) [inline xml-parsing/src/test/java/org/sonarsource/analyzer/commons/xml/checks/SimpleXPathBasedCheckTest.java]: I prefer to move this line out of this method, we don't need `scanFile` scope for it.

File: xml-parsing/src/test/resources/checks/SimpleXPathBasedCheck/xPathFailure.xml [hunk 9]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<!DOCTYPE foo [<!ELEMENT foo ANY >
+        <!ENTITY xxe SYSTEM "file:///dev/random" >]>
+
+<foo>&xxe;</foo>
```
Comment (vilchik-elena) [inline xml-parsing/src/test/resources/checks/SimpleXPathBasedCheck/xPathFailure.xml]: it  will be nice to add comment here about how this file fails and that we don't know why.

File: xml-parsing/src/test/java/org/sonarsource/analyzer/commons/xml/checks/SimpleXPathBasedCheckTest.java (hunk 8 above)
Comment (m-g-sonar) [reply xml-parsing/src/test/java/org/sonarsource/analyzer/commons/xml/checks/SimpleXPathBasedCheckTest.java]: In fact I even moved `asList` to `XmlFile` class

**Pull request merged**